docs/standard/garbage-collection/implementing-disposeasync.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Implement a DisposeAsync method
+description: 
+ms.date: 05/29/2020
+ms.technology: dotnet-standard
+dev_langs:
+  - "csharp"
+helpviewer_keywords:
+  - "DisposeAsync method"
+  - "garbage collection, DisposeAsync method"
+---
+
+# Implement a DisposeAsync method
+
+The <xref:System.IAsyncDisposable> interface was introduced as part of C# 8.0. You implement the <xref:System.IAsyncDisposable.DisposeAsync?displayProperty=nameWithType> method when you need to perform resource cleanup, just as you would when [implementing a Dispose method](implementing-dispose.md). One of the key differences however, is that this implementation allows for asynchronous cleanup operations. The <xref:System.IAsyncDisposable.DisposeAsync> returns a <xref:System.Threading.Tasks.ValueTask> that represents the asynchronous dispose operation.
+
+It is typical that when implementing the <xref:System.IAsyncDisposable> interface, classes will also implement the <xref:System.IDisposable> interface. All of the guidance for implementing the dispose pattern is also relevant to the asynchronous implementation. This article assumes that you're already familiar with how to [implement a Dispose method](implementing-dispose.md).
+
+## DisposeAsync() and DisposeAsyncCore()
+
+The <xref:System.IAsyncDisposable> interface requires the implementation of a single parameterless method, <xref:System.IAsyncDisposable.DisposeAsync>. Also, any non-sealed class should have an additional `DisposeAsyncCore()` method that also returns a <xref:System.Threading.Tasks.ValueTask>.
+
+- A `public` <xref:System.IAsyncDisposable.DisposeAsync?displayProperty=nameWithType> implementation that has no parameters.
+- A `protected virtual ValueTask DisposeAsyncCore()` method whose signature is:
+
+```csharp
+protected virtual ValueTask DisposeAsyncCore()
+{
+}
+```
+
+The `DisposeAsyncCore()` method is `virtual` so that derived classes can define additional cleanup in their overrides.
+
+### The DisposeAsync() method
+
+The `public` parameterless `DisposeAsync()` method is called implicitly in an `await using` statement, and its purpose is to free unmanaged resources, perform general cleanup, and to indicate that the finalizer, if one is present, doesn't have to run. Freeing the actual memory associated with a managed object is always the domain of the [garbage collector](index.md). Because of this, it has a standard implementation:
+
+```csharp
+public async ValueTask DisposeAsync()
+{
+    // Perform async cleanup.
+    await DisposeAsyncCore();
+
+    // Dispose of managed resources.
+    Dispose(false);
+    // Suppress finalization.
+    GC.SuppressFinalize(this);
+}
+```
+
+> [!NOTE]
```

Comment:
This paragraph was confusing to me. It took several reads to get to my current understanding (which may be wrong). Here's what I think you're trying to say. It matches my code comments below:

1. If a type implements `IAsyncDisposable`, it likely also implements `IDisposable`. 
1. `DisposeAsyncCore` disposes managed resources *asynchronously*, so you don't want to dispose of them *synchronously* as well. Therefore, call `Dispose(false)` instead of `Dispose(true)`.
1. The developers that use your type may dispose of it synchronously or asynchronously. Be prepared for either. (See my comments on the code)

I may be mis-understanding this, but that's what I got from this.

Reply:
These are all correct statements. I'll try to rephrase this more appropriately.